Fix JSON block model UV import

BolWaffy opened the pull request:
Overhauled the minecraft JSON block model importer so all imported models now have correct and consistent UV mapping.

Previously, certain models like Cake, Hoppers, Cauldrons, Scaffolding, and Composters had broken or misaligned UVs. Some other blocks also had alignment issues, especially with custom or non-standard JSON models.

This update adds an automatic UV generation and correction system that makes texture mapping consistent for all current and future JSON block models, eliminating the need for manual UV fixes or model-specific patches.

BolWaffy committed:
Fix JSON block model UV import

MCprep's JSON block model importer had inconsistent UV alignment and
broken UVs on some models, especially custom ones. This update adds
automatic UV generation and correction that provides consistent texture
mapping for all current and future JSON block models. The change
improves reliability of JSON imports and removes the need for manual
per-model UV patches.

TheDuckCow commented:
Thanks a ton for taking on this long-standing issue! I'll put some time aside to review it this weekend and let you know if I have any concerns, tagging @StandingPadAnimations in case you want to check in on it as well.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -331,6 +331,8 @@ def add_model(
 					obj_mats.append(mat)
 					materials.append(f"#{img}")
 
+	face_dir = ["north", "south", "up", "down", "west", "east"]
```
I would make this a tuple and place this as a constant at the top of the file (so all caps), since it's not mutated at any point in the operator.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +364,129 @@ def add_model(
-			face = bm.faces.new(
-				(verts[f[0]], verts[f[1]], verts[f[2]], verts[f[3]])
-			)
+				# Auto-calculate proportional UVs (0-16 scale) based on element bounds
```
It would be nice to have a basic overview of the algorithm used here, if possible

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +364,129 @@ def add_model(
+				)
+			except ValueError:
+				# If face cannot be created, skip it
+				continue
```
I would raise a warning here just so the user is aware

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +364,129 @@ def add_model(
+				uvs_base = [p_BR, p_BL, p_TL, p_TR]
+			elif face_name == "west": # X- (V-flip)
+				uvs_base = [p_TL, p_TR, p_BR, p_BL]
+			else:
+				# Default fallback (should not happen, just in case though)
+				uvs_base = [p_TL, p_TR, p_BR, p_BL]
```
If this shouldn't happen under normal circumstances, a warning should be raised so the end user is at least aware when this happens

TheDuckCow replied:
Added this to a commit I'm staging locally.

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +364,129 @@ def add_model(
+
+			# Assign the final computed UVs to the face loops
+			for j, loop in enumerate(face.loops):
+				loop[uv_layer].uv = uvs_final[j % len(uvs_final)]
```
A bounds check should be added here to avoid any unhandled exceptions

StandingPadAnimations replied:
While a bounds check on `j` was added, there is no bounds check for `j % len(uvs_final)`, nor for `uv_layer` when indexing into `loop` (unless I missed an earlier bounds check for `loop` somewhere)

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +364,129 @@ def add_model(
+				bmesh.ops.translate(bm, verts=face.verts, vec=0.0025 * face.normal)
+
+			# Apply rotation shift (uv_idx)
+			uvs_rot = [uvs_base[(k + uv_idx) % len(uvs_base)] for k in range(len(uvs_base))]
```
IMO this should be expanded to a for loop with a bounds check, just to avoid any unhandled exceptions related to going out of bounds

StandingPadAnimations replied:
There should still be a bounds check here

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +364,129 @@ def add_model(
+			try:
+				face = bm.faces.new(
+					(verts[f_for_face[0]], verts[f_for_face[1]], verts[f_for_face[2]], verts[f_for_face[3]])
+				)
```
A length check should be added here, to make sure `f_for_face` indeed has 4 elements

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +364,129 @@ def add_model(
+			p_BL = (u_min, v_min) # Bottom Left
+
+			# Face-specific base ordering (maps Tex-corners -> face loop positions)
+			if face_name == "up": # Y+ (V-flip and U-flip)
```
I've noticed a lot of these strings related to direction. Since they're reused constantly, I would suggest creating some constants at the top of the file to avoid typos, such as:

```py
UP_DIR = "up"
DOWN_DIR = "down"
NORTH_DIR = "north"
# So on and so forth...
```

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +364,129 @@ def add_model(
-			)
+				# Auto-calculate proportional UVs (0-16 scale) based on element bounds
+				f_bounds = e['from']  # [x1, y1, z1]
+				t_bounds = e['to']	  # [x2, y2, z2]
```
This should have a check to make sure `e` does indeed contain `from` and `to` at runtime, and raise an error if they don't. Also would be a good time to validate `f_bounds` and `t_bounds` to make sure that indeed they are lists of 3 points

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +364,129 @@ def add_model(
+			u_min = uv_coords[0] / 16.0
+			u_max = uv_coords[2] / 16.0
+			v_max = 1.0 - (uv_coords[1] / 16.0)  # Blender Top V
+			v_min = 1.0 - (uv_coords[3] / 16.0)  # Blender Bottom V
```
More of a stylistic comment, but `1.0` and `16.0` can simply be written as `1` and `16`, Python 3 automatically converts to floating point where needed

StandingPadAnimations reviewed (changes requested):
Have a couple of stylistic suggestions, as well as some related to checking bounds in places where bounds checks aren't currently performed

BolWaffy committed:
Code review feedback: add bound checks, constants, & stylistic fixes

Implemented the requested improvements:
- Converted `face_dir` to an uppercase tuple constant.
- Added comment documentation summarizing the UV generation algorithm/process.
- Added warnings where unexpected behavior occurs (using a mix of env.log() and raise warning)
- Simplified float literals
- Added bounds and length checks to prevent unhandled exceptions

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +386,153 @@ def add_model(
+				env.log(f"Face vertex index list expected 4 elements, but got {len(f_for_face)}. Skipping face.")
+				continue
+
+			try:
+				face_verts = []
+				for vert_idx in f_for_face:
+					if vert_idx < 0 or vert_idx >= len(verts):
+						env.log(f"Vertex index {vert_idx} is out of bounds (0 to {len(verts)-1}). Skipping face.")
+						raise IndexError("Vertex index out of bounds")
+					face_verts.append(verts[vert_idx])
+
+				face = bm.faces.new(tuple(face_verts))
+			except ValueError as e:
+				# If face cannot be created, skip it
+				env.log(f"Failed to create BMesh face: {e}. Skipping.")
+				continue
+			except IndexError:
+				# Catch the custom index error from the bounds check
+				continue
```
I'll defer to @TheDuckCow for a second opinion on if this could be simplified (while retaining the same logic of moving on to the next element if the index is out of bounds). I suspect it can't though, which is a bit annoying given how weird this is.

(Really wish Python had `goto` statements for weird situations like this)

TheDuckCow replied:
Pushing a change shortly - I believe this is functionally equivalent, although an extra line or two longer, has less indentation and more narrow try/exception handling (also slightly simplifies the comparison range, neat python feature that's useful in cases like this)

```python
			face_verts = []
			out_of_bounds = False
			for vert_idx in f_for_face:
				if 0 <= vert_idx < len(verts):
					face_verts.append(verts[vert_idx])
				else:
					env.log(f"Vertex index {vert_idx} is out of bounds (0 to {len(verts) - 1}). Skipping face.")
					out_of_bounds = True
					break  # No need to keep checking this face
			if out_of_bounds:
				continue
			try:
				face = bm.faces.new(tuple(face_verts))
			except ValueError as e:
				env.log(f"Failed to create BMesh face: {e}. Skipping.")
				continue
```

StandingPadAnimations commented on MCprep_addon/spawner/mcmodel.py:
```diff
@@ -363,55 +386,153 @@ def add_model(
+				uvs_final = uvs_rot
+
+			# For cardinal faces (north/south/east/west) mirror over both axes
+			if face_name in {NORTH_DIR, SOUTH_DIR, EAST_DIR, WEST_DIR}:
```
This can be done as a tuple instead of a dictionary


```suggestion
			if face_name in (NORTH_DIR, SOUTH_DIR, EAST_DIR, WEST_DIR):
```

StandingPadAnimations reviewed (commented):
Requesting some more changes, mainly to do with some small things. I'm also interested in seeing a demonstration of this, though I'll probably be able to test this locally on Sunday

BolWaffy committed:
Added more fail-safes, changed Array to Tuple

Added more bounds checks, edited to Tuple, more fail-safes

StandingPadAnimations committed:
style: Make JSON UV algorithm overview comment not overflow

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

StandingPadAnimations committed:
refactor: Simplify sections in UV algorithm

Signed-off-by: Maryam Sheikh <mahid@standingpad.org>

TheDuckCow commented:
Thanks for flagging, will be looking tonight or tomorrow morning. I'll check in on the test runner failure as well

BolWaffy committed:
Fixed Error: TypeError: argument of type 'BMLoop' is not iterable

Found and fixed the problem.

TheDuckCow commented on MCprep_addon/spawner/mcmodel.py:
Other audit notes: that qa script actually came in handy. Here's the current total list of files that fail to import:

```
Those that failed:
black_shulker_box: Error: JSON model does not contain any actual geometry

blast_furnace: Not enough geo generated
blast_furnace_on: Not enough geo generated
block: Error: JSON model does not contain any actual geometry

blue_shulker_box: Error: JSON model does not contain any actual geometry

brown_shulker_box: Error: JSON model does not contain any actual geometry

carved_pumpkin: Not enough geo generated
copper_chest: Error: JSON model does not contain any actual geometry

copper_golem_statue: Error: JSON model does not contain any actual geometry

cube: Not enough geo generated
cyan_shulker_box: Error: JSON model does not contain any actual geometry

decorated_pot: Error: JSON model does not contain any actual geometry

dispenser: Not enough geo generated
dried_ghast: Not enough geo generated
dropper: Not enough geo generated
end_gateway: Error: JSON model does not contain any actual geometry

exposed_copper_chest: Error: JSON model does not contain any actual geometry

exposed_copper_golem_statue: Error: JSON model does not contain any actual geometry

furnace: Not enough geo generated
furnace_on: Not enough geo generated
gray_shulker_box: Error: JSON model does not contain any actual geometry

green_shulker_box: Error: JSON model does not contain any actual geometry

jack_o_lantern: Not enough geo generated
lava: Error: JSON model does not contain any actual geometry

light_00: Error: JSON model does not contain any actual geometry

light_01: Error: JSON model does not contain any actual geometry

light_02: Error: JSON model does not contain any actual geometry

light_03: Error: JSON model does not contain any actual geometry

light_04: Error: JSON model does not contain any actual geometry

light_05: Error: JSON model does not contain any actual geometry

light_06: Error: JSON model does not contain any actual geometry

light_07: Error: JSON model does not contain any actual geometry

light_08: Error: JSON model does not contain any actual geometry

light_09: Error: JSON model does not contain any actual geometry

light_blue_shulker_box: Error: JSON model does not contain any actual geometry

light_gray_shulker_box: Error: JSON model does not contain any actual geometry

lime_shulker_box: Error: JSON model does not contain any actual geometry

magenta_shulker_box: Error: JSON model does not contain any actual geometry

orange_shulker_box: Error: JSON model does not contain any actual geometry

oxidized_copper_chest: Error: JSON model does not contain any actual geometry

oxidized_copper_golem_statue: Error: JSON model does not contain any actual geometry

pink_shulker_box: Error: JSON model does not contain any actual geometry

pitcher_crop_top_stage_0: Error: JSON model does not contain any actual geometry

pitcher_crop_top_stage_1: Error: JSON model does not contain any actual geometry

pitcher_crop_top_stage_2: Error: JSON model does not contain any actual geometry

purple_shulker_box: Error: JSON model does not contain any actual geometry

red_shulker_box: Error: JSON model does not contain any actual geometry

shulker_box: Error: JSON model does not contain any actual geometry

trapped_chest: Error: JSON model does not contain any actual geometry

weathered_copper_chest: Error: JSON model does not contain any actual geometry

weathered_copper_golem_statue: Error: JSON model does not contain any actual geometry

white_shulker_box: Error: JSON model does not contain any actual geometry

yellow_shulker_box: Error: JSON model does not contain any actual geometry
```

And for fun, the scene output. Currently zooming around seeing if anything looks off beyond the errors already listed.

<img width="1003" height="746" alt="Screenshot 2025-11-11 at 10 08 43 AM" src="https://github.com/user-attachments/assets/48a1ec67-8a35-44ee-b7b8-bb1a80f61fbe" />

<img width="920" height="746" alt="Screenshot 2025-11-11 at 10 15 29 AM" src="https://github.com/user-attachments/assets/feb14aef-f030-4b1f-9d6a-ece4ffd67a17" />


The only ones visually looking off are white with no textures: fence_inventory, cross_emissive, cross, crop, coral_fan, coral_wall_fan, pressure_plate_down, pressure_plate_up, tinted_cross.. and some more -> All expected, as these are more "base shapes" and tbh we should find a way to filter them out (ideally procedurally)

And of course, the more self evident fact that _many_ of these models are meant to be really instanced with sets of other models, most notably doors. And doesn't account for anything else that might be missing like beds.

TheDuckCow reviewed (approved):
@BolWaffy can you grant me write access to your repo so I can push a change I've committed locally? With that, we'll be ready to merge I think (worst case, I'll push it after merging your branch, mostly just cleanup)

StandingPadAnimations commented:
> @BolWaffy can you grant me write access to your repo so I can push a change I've committed locally? With that, we'll be ready to merge I think (worst case, I'll push it after merging your branch, mostly just cleanup)

The PR has enabled the option for maintainers to edit, so you should be able to just push the changes to the branch (at least that works on my end)

TheDuckCow commented:
Yeah not really sure why it's blocking me out, may have something to do with that I renamed the local target of the branch name to not conflict with my own dev branch (was creating auto merges).

```
mcprep $ gh pr checkout 688 --branch pr-688 
...
mcprep $ git push git@github.com:BolWaffy/MCprep.git HEAD:dev
error: Authentication error: Authentication required: You must have push access to verify locks
error: failed to push some refs to 'github.com:BolWaffy/MCprep.git'
```

I'll just merge this locally and cherry pick my change, I need to make another commit for the version update anyways.
